Create Initial API Documentation (#19)
icns-xyz/icns: docs/API.md

# ICNS API Documentation

Introduction
------------

Welcome to the ICNS API documentation! The ICNS API allows you to interact with the Interchain Nameservice, a decentralized protocol that enables users to register and manage human-readable names on the Interchain.

Contract Address
----------------

*   Name NFT

>     *   Mainnet: TBD

Reviewer: ```suggestion
    *   Mainnet:  `osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd`
```

Reviewer 2: Sunny got the Name NFT & Resolver contracts flip-flopped

![image](https://user-images.githubusercontent.com/38844339/215022796-e2730892-d79c-4736-8b68-6ea8393fe2d1.png)

```suggestion
    *   Mainnet: `osmo1mypljhatv0prfr9cjzzvamxdf2ctg34xkt50sudxads9zhqnyneqjuvy26`
```

*   Resolver

>     *   Mainnet: TBD

Reviewer: ```suggestion
    *   Mainnet:  `osmo1mypljhatv0prfr9cjzzvamxdf2ctg34xkt50sudxads9zhqnyneqjuvy26`
```

Reviewer 2: Sunny got the Name NFT & Resolver contracts flip-flopped

![image](https://user-images.githubusercontent.com/38844339/215022770-5e47142a-3904-4a2f-898d-fec088f8005c.png)

```suggestion
    *   Mainnet: `osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd`
```

*   Registrar

>     *   Mainnet: TBD

Reviewer: ```suggestion
    *   Mainnet:  `osmo1llh07xn7pcst3jqm0xpsucf90lzugfskkkhk8a3u2yznqmse4l5smydwpw`
```

API Endpoints
-------------

### Resolver APIs

*   Query - config
    *   Gets the config of the resolver contract, including the Name-NFT contract address.
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "config": { }})}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "config": { }})}
```

    ```
    
*   Query - addresses\_response
    *   Returns a list of (bech32\_prefix, bech32\_address) associated with the given ICNS name
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "addresses": { "name": "{icns_name}" })}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "addresses": { "name": "{icns_name}" })}
```

    ```
    
*   Query - address
    *   Gets the address associated with the given an ICNS name
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "icns_names": { "address": "{bech32Address}" }})} 

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address": { "name": "{icns_name}", "bech32_prefix": "{bech32_prefix}" }})} 
```

    ```
    
    *   Note: you can register multiple addresses to a name. check for `name` and `primary_name`
    *   Method to modify the `primary_name` is implemented on the contract, but not on the frontend (future feature)
*   Query - names
    *   Returns a list of names associated with the given address
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "names": { "{bech32Address}" }})}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "names": { "{bech32Address}" }})}
```

    ```
    
    *   NOTE: this query only returns name itself, not full icns name
        *   does not include ICNS bech32 suffix
*   Query - icns\_names
    *   Returns a list of all ICNS names and bech32 prefixes associated with the given address
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "icns_names": { "{bech32Address}" }})}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "icns_names": { "{bech32Address}" }})}
```

    ```
    
*   Query - primary\_name
    *   Returns the primary ICNS associated with the given address
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "primary_name": { "{bech32Address}" }})}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "primary_name": { "address": "{bech32Address}" }})}
```

    ```
    
*   Query - admin
    *   Returns a list of admins queried from the Name NFT contract.
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "admin": { }})}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "admin": { }})}
```

    ```
    
*   Query - address\_by\_icns
    *   Returns the bech32 addresses for the associated full (including the bech32 prefix) ICNS name.
    
    ```plain

>     {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}

Reviewer 2: ```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address_by_icns": { "icns": "{twitter_username}.{bech32_prefix}" }})}
```

    ```
    
*   Execute Message - claim
*   Execute Message - update\_verifier\_pub\_keys
*   Execute Message - set\_verification\_threshold
*   Execute Message - set\_name\_nft\_address
*   Execute Message - set\_minting\_fee
*   Execute Message - withdraw\_funds

### **Name NFT API**

*   Query - admin
    *   Returns the admin address
*   Query - is\_admin
    *   Checks whether the specified address is an admin
*   Query - transferrable
    *   Returns whether Name NFTs are currently transferrable
*   Query - owner\_of
    *   Check the owner of the given token ID (optional filter on including expired owners)
*   Query - approval
    *   Check for the approval response for a given token ID and spender (optional filter on including expired owners)
*   Query - approvals
    *   Check for the approval response for a given token ID (optional filter on including expired owners)
*   Query - all\_operators
*   Query - num\_tokens
*   Query - contract\_info
*   Query - nft\_info
*   Query - all\_nft\_info
*   Query - tokens
*   Query - all\_tokens
*   Query - minter

Examples
--------

**Get a bech32 address from an ICNS name**

```plain

> {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}

Reviewer 2: Dashes to underscores.
```suggestion
{endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address_by_icns": { "icns": {twitter_username}.{bech32_prefix} }})}
```

```

> *   Example: [https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K](https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K)

Reviewer: ```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K](https://lcd.osmosis.zone//cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K)
```

Reviewer 2: In addition to the endpoint update to mainnet, the Contract address needs updated to mainnet version and the query was an `icns_names` one so updated with a `address_by_icns` one.
```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJhZGRyZXNzX2J5X2ljbnMiOiB7ImljbnMiOiAiTm9zTm9kZV8ub3NtbyJ9fQo=](https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJhZGRyZXNzX2J5X2ljbnMiOiB7ImljbnMiOiAiTm9zTm9kZV8ub3NtbyJ9fQo=)
```

*   Response

```plain
{
  "data": {"names":["TonyYun9.osmo"],"primary_name":"TonyYun9.osmo"}
}
```

  

> **Get a ICNS name of a bech32 address**

Reviewer 2: ```suggestion
**Get all ICNS names associated with a bech32 address**
```

```plain

> {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}

Reviewer 2: ```suggestion
{endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "names": { "{bech32Address}" }})}
```

```

> *   Example: [https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=](https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=)

Reviewer: ```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=](https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=)
```

Reviewer 2: In addition to the endpoint update to mainnet, the Contract address needs updated to mainnet version and the query was an `icns_names` one so updated with a `names` one.
```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJuYW1lcyI6IHsiYWRkcmVzcyI6ICJvc21vMTBqbThmdmR5cWxqNzh3MGo1bmF3Yzc2d3NuNHBxbWR4ajRxNXpsIn19Cg==](https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJuYW1lcyI6IHsiYWRkcmVzcyI6ICJvc21vMTBqbThmdmR5cWxqNzh3MGo1bmF3Yzc2d3NuNHBxbWR4ajRxNXpsIn19Cg==)
```

*   Response

```plain
{

>   "data": {"address":"osmo1vv6hruquzpty4xpks9znkw8gys5x4nsng4kery"}

Reviewer 2: Updating the response to reflect the new example in https://github.com/icns-xyz/icns/pull/19#discussion_r1089647680
```suggestion
  "data": {"names": ["NosNode_", "NosNode_Intern"], "primary_name": "NosNode_"}
```
